rest_framework/fields.py

```diff
@@ -653,6 +654,30 @@ def to_representation(self, value):
         return str(value)
 
 
+class IPAddressField(CharField):
+    """Support both IPAddressField and GenericIPAddressField"""
+
+    default_error_messages = {
+        'invalid': _('Enter a valid IPv4 or IPv6 address.'),
+    }
+
+    def __init__(self, protocol='both', unpack_ipv4=False, **kwargs):
+        self.protocol = protocol
+        self.unpack_ipv4 = unpack_ipv4
+        super(IPAddressField, self).__init__(**kwargs)
+        validators, error_message = ip_address_validators(protocol, unpack_ipv4)
+        self.validators.extend(validators)
+
+    def to_internal_value(self, data):
+        if data and ':' in data:
+            try:
+                return clean_ipv6_address(data, self.unpack_ipv4)
+            except DjangoValidationError:
+                self.fail('invalid', value=data)
```

Comment:
Hi Tom,

I'm away from my computer for few days so I will be able to reply this more properly at the end of the week. Basically I noticed that ipv4_unpack was needed anyway by the method that returns validator, that's why I left it. I need to check if tests still pass always defaulting it to True when is both. It doesn't have to be always True, nor always false.

Sent from my Windows Phone

-----Original Message-----
From: "Tom Christie" <notifications@github.com>
Sent: ‎23/‎03/‎2015 10:59
To: "tomchristie/django-rest-framework" <django-rest-framework@noreply.github.com>
Cc: "Andrea Grandi" <a.grandi@gmail.com>
Subject: Re: [django-rest-framework] IPAddressField and tests imporvements(#2747)

In rest_framework/fields.py:
> +        'invalid': _('Enter a valid IPv4 or IPv6 address.'),
> +    }
> +
> +    def __init__(self, protocol='both', unpack_ipv4=False, **kwargs):
> +        self.protocol = protocol
> +        self.unpack_ipv4 = unpack_ipv4
> +        super(IPAddressField, self).__init__(**kwargs)
> +        validators, error_message = ip_address_validators(protocol, unpack_ipv4)
> +        self.validators.extend(validators)
> +
> +    def to_internal_value(self, data):
> +        if data and ':' in data:
> +            try:
> +                return clean_ipv6_address(data, self.unpack_ipv4)
> +            except DjangoValidationError:
> +                self.fail('invalid', value=data)
Looking good. I think we arrived at a slight tweak to this, that involved us getting rid of the unpack_ipv4 flag and just always having that as True for 'both'. Something like this?...
if self.protocol in ('both', 'ipv6'):
    unpack_ipv4 = (self.protocol == 'both')
    return clean_ipv6_address(data)
return data
—
Reply to this email directly or view it on GitHub.